src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,666 @@
+): {
+	spansById: Map<number, SpanWithChildren>;
+	parsedSpans: SpanWithChildren[];
+	executeBlockSpanId: number;
+} {
+	// IMPORTANT for downstream logic: Make sure spans are sorted.
+	spans.sort((a, b) => a.id - b.id);
+
+	const spansById = new Map<number, SpanWithChildren>();
+	const parsedSpans = [];
+	let executeBlockSpanId;
+	for (const span of spans) {
+		const spanWithChildren = { ...span, children: [] };
+
+		if (
+			span.name === SPANS.executeBlock.name &&
+			span.target === SPANS.executeBlock.target
+		) {
+			executeBlockSpanId = span.id;
+		}
+
+		parsedSpans.push(spanWithChildren);
+		spansById.set(span.id, spanWithChildren);
+	}
+
+	if (executeBlockSpanId === undefined) {
+		throw new InternalServerError('execute_block span could not be found');
+	}
+
+	// Go through each span, and add its Id to the `children` array of its
+	// parent span, creating tree of spans starting with `execute_block` span as
+	// the root.
+	for (const span of spans) {
+		if (!span.parentId) {
+			continue;
+		}
+
+		if (!spansById.has(span.parentId)) {
+			throw new InternalServerError(
+				'Expected spans parent to exist in spansById'
+			);
+		}
+
+		spansById.get(span.parentId)?.children.push(span.id);
+	}
+
+	return { spansById, parsedSpans, executeBlockSpanId };
+}
+
+/**
+ * Find the Ids of al the descendant spans of `root`.
+ *
+ * @param root span which we want all the descendants of
+ * @param spansById map of span id => span with children
+ */
+function findDescendants(
+	root: number,
+	spansById: Map<number, SpanWithChildren>
+): number[] {
+	// Note: traversal order doesn't matter here
+	const stack = spansById.get(root)?.children;
+	if (!stack) {
+		return [];
+	}
+
+	const descendants = [];
+	while (stack.length) {
+		const curId = stack.pop();
+		if (!curId) {
+			break;
+		}
+
+		descendants.push(curId);
+
+		const curSpanChildren = spansById.get(curId)?.children;
+		curSpanChildren && stack.push(...curSpanChildren);
+	}
+
+	return descendants;
+}
+
+/**
+ * Extract extrinsic indexes from all `:extrinsic_index` storage item read events.
+ *
+ * @param spanIds List of spanIds for an apply extrinsic action group
+ * @param extrinsicIndexBySpanId Map of spanId => extrinsicIndex
+ */
+function extractExtrinsicIndex(
+	spanIds: number[],
+	extrinsicIndexBySpanId: Map<number, BN>
+): BN {
+	// There are typically multiple reads of extrinsic index when applying an extrinsic,
+	// So we get all of those reads, which should all be the same number.
+	const extrinsicIndex = spanIds.reduce((acc, id) => {
+		if (!extrinsicIndexBySpanId.has(id)) {
+			return acc;
+		}
+
+		acc.push(extrinsicIndexBySpanId.get(id) as BN);
+		return acc;
+	}, [] as BN[]);
+
+	if (extrinsicIndex.length === 0) {
+		// Since we assume the `spanIds` are for span from an apply extrinsic action group,
+		// we assume there is always an `:extrinsic_index` event
+		throw new InternalServerError('Expected at least one extrinsic index');
+	}
+
+	if (extrinsicIndex.length > 1) {
```

Comment:
Interesting sanity check. Could an issue like the originate from the substrate rpc client, or would it always be the logic in this endpoints scope?

Reply:
Kind of both. If the substrate logic changes this may need to change. On the other hand if I am making faulty assumptions (namely that last read to an extrinsic index within an action group will always be for the current extrinsic index and not the next or former) the way we get the extrinsic index would need to change.